### PR TITLE
Methods

### PR DESCRIPTION
**Pull request by CalyCherkaoui**
The objective of this exercise is to learn about Enumerable methods in Ruby by coding them from scratch.
The methods built in this exercise are:
- my_each that mimics the ruby built-in method "each"
- my_each_with_index the ruby built-in method "each_with_index"
- my_select that mimics the ruby built-in method "select"
- my_all? that mimics the ruby built-in method "all?"
- my_any? that mimics the ruby built-in method "any?"
- my_none? that mimics the ruby built-in method "none?"
- my_count that mimics the ruby built-in method "count"
- my_map that mimics the ruby built-in method "map"
- my_inject that mimics the ruby built-in method "inject"

### CODE REVIEW

File: method.rb [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+    i = 0
+    while i < size
+      yield self[i], i
+      i += 1
+    end
+    self
+  end
+
+  def my_select
+    return 'no block given' unless block_given?
```
Comment (Forison) [inline method.rb]: should have working `my_select`  identical and returns the same thing as ruby's [`select`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-select). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_select.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

1. ```p [1,2,3,4,5].my_select{|a| a.even? }``` must return same result as```p [1,2,3,4,5].select{|a| a.even? }``` 
2. ```[1,2,3,4,5].my_select{|a| p a.even? }``` must return same result as
``` [1,2,3,4,5].select{|a| p a.even? }```

File: method.rb [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+    i = 0
+    while i < size
+      output_array << self[i] if (yield i) == true
+      i += 1
+    end
+    output_array
+  end
+
+  def my_all?
+    return 'no block given' unless block_given?
```
Comment (Forison) [inline method.rb]: * [x] There should be a working `my_all?`  identical and returns the same thing as ruby's [`all?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-all-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_all.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
 #### compare the following examples by replacing the custom method with the original method in each example and rectify the error that arises.
```p %w[ant bear cat].my_all? { |word| word.length >= 3 }```   must return```true```

```p %w[ant bear cat].my_all? { |word| word.length >= 4 }```   must return```false```

```p %w[ant bear cat].my_all?(/t/)````   must return``` false```

```p [1, 2i, 3.14].my_all?(Numeric) ```   must return``` true```

```p [nil, true, 99].my_all?```   must return``` false```

```p [].my_all? ```   must return```true```

File: method.rb [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+      status_cumulator &&= yield i
+      i += 1
+      break if i == size || status_cumulator == false
+    end
+    status_cumulator
+  end
+
+  def my_any?
+    return 'no block given' unless block_given?
+
```
Comment (Forison) [inline method.rb]: * [x] There should be a working `my_any?`  identical and returns the same thing as ruby's [`any?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-any-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_any.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
#### compare the following examples by replacing the custom method with the original method in each example and rectify the error that arises.
```p %w[ant bear cat].my_any? { |word| word.length >= 3 } ```   must return```  true```

```p %w[ant bear cat].my_any? { |word| word.length >= 4 } ```   must return```  true```

```p %w[ant bear cat].my_any?(/d/) ```   must return```  false```

```p [nil, true, 99].my_any?(Integer) ```   must return```  true```

```p [nil, true, 99].my_any? ```   must return```  true```

```p [].my_any?```   must return```  false```

File: method.rb [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+    loop do
+      status_cumulator ||= yield i
+      i += 1
+      break if i == size || status_cumulator == true
+    end
+    status_cumulator
+  end
+
+  def my_none?
+    return 'no block given' unless block_given?
```
Comment (Forison) [inline method.rb]: * [x] should be a working `my_none?` identical and returns the same thing as ruby's [`none?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-none-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_none.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
#### compare the following examples by replacing the custom method with the original method in each example and rectify the error that arises.
```p %w{ant bear cat}.my_none? { |word| word.length == 5 } ```   must return``` true```

```p %w{ant bear cat}.my_none? { |word| word.length >= 4 } ```   must return``` false```

```p %w{ant bear cat}.my_none?(/d/) ```   must return```true```

```p [1, 3.14, 42].my_none?(Float) ```   must return``` false```

```p [].my_none?```   must return``` true```

```p [nil].my_none? ```   must return``` true```

```p [nil, false].my_none?  ```   must return``` true```

```p [nil, false, true].my_none? ```   must return``` false```

File: method.rb [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+    loop do
+      status_cumulator &&= yield i
+      i += 1
+      break if i == size || status_cumulator == false
+    end
+    !status_cumulator
+  end
+
+  def my_count
+    return size unless block_given?
```
Comment (Forison) [inline method.rb]: * [x] There should be a working `my_count` identical and returns the same thing as ruby's [`count`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-count). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_count.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- ```p [1, 2, 4, 2].my_count(2)``` should return  same result as ```p [1, 2, 4, 2].count(2)```

File: method.rb [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+    i = 0
+    while i < size
+      counter += 1 if yield self[i]
+      i += 1
+    end
+    counter
+  end
+
+  def my_inject
+    return 'no block given' unless block_given?
```
Comment (Forison) [inline method.rb]: * [x] There should be a working `my_inject`  identical and returns the same thing as ruby's [`inject`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-inject). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_inject.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
> note my_inject should throw an error when no block is given
1. ```p (5..10).my_inject { |sum, n| sum + n }``` should return  ```45```
2. ```p (5..10).my_inject(1) { |product, n| product * n }``` should return ```151200```

**Review by Forison — CHANGES_REQUESTED** [review]
## Status:: Changes required❌
# 🙋‍♂️ Houda!!!
#### You have done a great job so far, your pull request is quite detailed, however, here are a few changes we need to effect to get this project approved😎
* [x] Kindly separate your test examples from the actual code, you can place it in a separate file in the repo
* [x] The methods ```my_each```, ```my_each_with_index``` ```my_select``` and ```my_map``` should return an enumerator when no block is given.
``` return to_enum(:name_of_method_in_scope) unless block_given? ``` 
instead of
 ``` return 'no block given' unless block_given? ```

### Kindly check my comments 👇below👇 and address the issues raised thereof.
## That is it, kindly make the changes and submit the project for a re-review😎
#### Happy coding💻
Addo

File: method.rb [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+
+    output_array = []
+    my_each do |item|
+      output_array << item if (yield item) == true
+    end
+    output_array
+  end
+
+  def my_all?(parameter = nil) # rubocop:disable Metrics/CyclomaticComplexity
+    status_cumulator = true
```
Comment (Forison) [inline method.rb:51]: * [x] There should be a working `my_all?`  identical and returns the same thing as ruby's [`all?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-all-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_all.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
 #### compare the following examples by replacing the custom method with the original method in each example and rectify the error that arises.
``` p [1, 2i, 3.14].my_all?(Numeric) #=> must return true```

``` p [nil, true, 99].my_all? #=>must return false```

``` p [].my_all? #=> true```

File: method.rb [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+      status_cumulator &&= (self[i] == parameter) if parameter
+      i += 1
+      break if i == size || status_cumulator == false
+    end
+    status_cumulator
+  end
+
+  def my_any?(parameter = nil) # rubocop:disable Metrics/CyclomaticComplexity
+    status_cumulator = false
+    i = 0
```
Comment (Forison) [inline method.rb:121]: * [x] There should be a working `my_any?`  identical and returns the same thing as ruby's [`any?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-any-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_any.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
#### compare the following examples by replacing the custom method with the original method in each example and rectify the error that arises.
```p [nil, true, 99].my_any?(Integer) #=>must return true```

```p [nil, true, 99].my_any? #=> must return true```

```p [].my_any? #=> must return false```

File: method.rb [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+      status_cumulator ||= self[i].is_a(parameter) if parameter.is_a?(Class)
+      status_cumulator ||= (self[i] == parameter) if parameter
+      i += 1
+      break if i == size || status_cumulator == true
+    end
+    status_cumulator
+  end
+
+  def my_none?(parameter = nil)
+    !my_any?(parameter)
```
Comment (Forison) [inline method.rb]: * [x] should be a working `my_none?` identical and returns the same thing as ruby's [`none?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-none-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_none.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
#### compare the following examples by replacing the custom method with the original method in each example and rectify the error that arises.
```p %w{ant bear cat}.my_none? { |word| word.length >= 4 } # must return false```

```p [ ].my_none? # must return true```

```p [nil, false, true].my_none? # must return false```

**Review by Forison — CHANGES_REQUESTED** [review]
## Status:: Changes required❌
# 🙋‍♂️ Houda!!!
#### Good job on addressing some of the issues raised in the previous review, here are a few other issues which were not addressed.
* [x] There should Be working ```multiply_els```  that uses ```my_inject``` as described on Odin 
([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/multiply_els.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- The multiply_els method is still missing, kindly add it and it should be outside the module
### Check comment below for other issues that are still problematic😥.
## That is it, kindly make the change and submit the project for a re-review😎.
#### Happy coding💻 
Addo Forison

File: method.rb [hunk 10]
```diff
@@ -0,0 +1,218 @@
...
+    end
+    output_array
+  end
+
+  def parameter_condition?(parameter)
+    parameter.is_a?(Regexp) || parameter.is_a?(Class) || parameter.nil?
+  end
+
+  def my_all?(parameter = nil) # rubocop:disable Metrics/CyclomaticComplexity, Metrics/PerceivedComplexity, Metrics/MethodLength, Metrics/AbcSize
+    status_cumulator = true
```
Comment (majovanilla) [inline method.rb]: - [x] should have working `my_all?` *identical and returns the same thing as ruby's [`all?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-all-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_all.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

- Your method doesn't consider the possibility when no block or argument is given. It should return true if the enumerable contains truthy values and false if it contains at least one falsy value.

- You clearly made an effort with your method, but it's extremely long and somewhat over-complicated. After you finish the assignment, you should try to refactor your code and make it simpler, as an exercise to write cleaner code. **Optional**

File: method.rb [hunk 11]
```diff
@@ -0,0 +1,218 @@
...
+        i += 1
+        break if i == size || status_cumulator == false
+      end
+    end
+    status_cumulator
+  end
+
+  def my_any?(parameter = nil) # rubocop:disable Metrics/CyclomaticComplexity, Metrics/PerceivedComplexity, Metrics/MethodLength, Metrics/AbcSize
+    status_cumulator = false
+    i = 0
```
Comment (majovanilla) [inline method.rb]: - [x] should have working `my_any?` identical and returns the same thing as ruby's [`any?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-any-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_any.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

- Your method doesn't consider the possibility when no block or argument is given. It should return true if the enumerable contains at least one truthy value.

- You clearly made an effort with your method, but it's extremely long and somewhat over-complicated. After you finish the assignment, you should try to refactor your code and make it simpler, as an exercise to write cleaner code. **Optional**

File: method.rb [hunk 12]
```diff
@@ -0,0 +1,218 @@
...
+        status_cumulator ||= (self[i] == parameter)
+        i += 1
+        break if i == size || status_cumulator == true
+      end
+    end
+    status_cumulator
+  end
+
+  def my_none?(parameter = nil)
+    !my_any?(parameter)
```
Comment (majovanilla) [inline method.rb]: I strongly advise you against this. Not only you are not making yourself think about the solution to this, but you are also depending on another method. This is not a good practice and you should refactor your code.

- Right now, you method returns false when given a truthy block of code. It should return false.
- Your method doesn't consider the possibility when no block or argument is given. It should return true if the enumerable contains no truthy value.

File: method.rb [hunk 13]
```diff
@@ -0,0 +1,218 @@
...
+    while i < size
+      cumulator = yield cumulator, output_array[i]
+      i += 1
+    end
+    cumulator
+  end
+
+  def my_map(&proc)
+    output_array = []
+    if block_given?
```
Comment (majovanilla) [inline method.rb]: - [x] should have working `my_map` **(covered by automated tests)** identical and returns the same thing as ruby's [`map`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-map). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_map.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

- Your method should return an enumerator if no block or argument is given.
- Your method should execute only the proc when both a block and a proc are given. Right now is raising an argument error.

File: method.rb [hunk 14]
```diff
@@ -0,0 +1,218 @@
...
+    else
+      cumulator = output_array[0]
+      i = 1
+    end
+    while i < size
+      cumulator = yield cumulator, output_array[i]
+      i += 1
+    end
+    cumulator
+  end
```
Comment (majovanilla) [inline method.rb]: - [ ] should have working `my_inject` identical and returns the same thing as ruby's [`inject`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-inject). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_inject.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

- Your method should work with symbols as arguments (:+) and perform the right operation.

**Review by majovanilla — CHANGES_REQUESTED** [review]
Hi @CalyCherkaoui 

## You did a *great* job implementing the changes on this project! 💜 

However, there are still some things you need to take care in some of your methods. 
Please check the commented code below.


## Please, make all REQUIRED changes and submit your request again! 😀

If you have any questions, feel free to send me a message through Slack (@majo)
I would be happy to help! 👩‍💻 


Happy coding 🤓

--Maria Reyes

File: method.rb (hunk 13 above)
Comment (CalyCherkaoui) [reply method.rb]: While running the original  ruby's method map with the same examples as my_map,  map returns this error when we pass it both a bloc and a proc:
"both block arg and actual block given".
my_map is also returning the same when I pass it both a bloc and a proc. Please, would you like to explain to me more about this point?

File: method.rb [hunk 15]
```diff
@@ -0,0 +1,248 @@
...
+      output_array << item if (yield item) == true
+    end
+    output_array
+  end
+
+  def parameter_condition?(parameter)
+    parameter.is_a?(Regexp) || parameter.is_a?(Class) || parameter.nil?
+  end
+
+  def my_all?(parameter = nil) # rubocop:disable Metrics/CyclomaticComplexity, Metrics/PerceivedComplexity, Metrics/MethodLength, Metrics/AbcSize
```
Comment (serdg0) [inline method.rb]: **my_all** when no block or argument is given it should return true when none of the collection members are false or nil

File: method.rb [hunk 16]
```diff
@@ -0,0 +1,248 @@
...
+      loop do
+        status_cumulator &&= (self[i] == parameter)
+        i += 1
+        break if i == size || status_cumulator == false
+      end
+    end
+    status_cumulator
+  end
+
+  def my_any?(parameter = nil) # rubocop:disable Metrics/CyclomaticComplexity, Metrics/PerceivedComplexity, Metrics/MethodLength, Metrics/AbcSize
```
Comment (serdg0) [inline method.rb]: **my_any** when no block or argument is given it should return true if at least one of the collection is not false or nil

File: method.rb [hunk 17]
```diff
@@ -0,0 +1,248 @@
...
+      loop do
+        status_cumulator ||= (self[i] == parameter)
+        i += 1
+        break if i == size || status_cumulator == true
+      end
+    end
+    status_cumulator
+  end
+
+  def my_none?(parameter = nil, &my_block)
```
Comment (serdg0) [inline method.rb:188]: **my_none** when no block or argument is given it should return true only if none of the collection members is true

**Review by serdg0 — CHANGES_REQUESTED** [review]
Hello @CalyCherkaoui !

You are almost there! :racing_car: 

You only need to fix considering the following edge case: When no block or argument is given, in the methods I show you below.

When done, kindly submit for another CRR.

**Happy Coding!** :computer: :nerd_face:
**_Sergio Diaz_**

**Review by rOluochKe — APPROVED** [review]
Hello Team
### You have done great and for also making the changes from the previous reviews :clap: The project looks good and therefore, it has been approved. Congratulations :boom: :100:
### You can merge and submit a project completion form from the dashboard. Good luck with the next project.
- Happy coding!

**Pull request merged**